<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Linkage Disequilibrium (LD)</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsLDpair {SNPRelate}"><tr><td>snpgdsLDpair {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Linkage Disequilibrium (LD)
</h2>

<h3>Description</h3>

<p>Return a LD value between snp1 and snp2.
</p>


<h3>Usage</h3>

<pre>
snpgdsLDpair(snp1, snp2, method = c("composite", "r", "dprime", "corr"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>snp1</code></td>
<td>
<p>a vector of SNP genotypes (0 &ndash; BB, 1 &ndash; AB, 2 &ndash; AA)</p>
</td></tr>
<tr valign="top"><td><code>snp2</code></td>
<td>
<p>a vector of SNP genotypes (0 &ndash; BB, 1 &ndash; AB, 2 &ndash; AA)</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
<p>&quot;composite&quot;, &quot;r&quot;, &quot;dprime&quot;, &quot;corr&quot;, see details</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Four methods can be used to calculate linkage disequilibrium values:
&quot;composite&quot; for LD composite measure, &quot;r&quot; for R coefficient (by EM algorithm
assuming HWE, it could be negative), &quot;dprime&quot; for D', and &quot;corr&quot; for
correlation coefficient. The method &quot;corr&quot; is equivalent to &quot;composite&quot;, when
SNP genotypes are coded as: 0 &ndash; BB, 1 &ndash; AB, 2 &ndash; AA.
</p>


<h3>Value</h3>

<p>Return a numeric vector:
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>ld</code></td>
<td>
<p>a measure of linkage disequilibrium</p>
</td></tr>
</table>
<p>if <code>method = "r" or "dprime"</code>,
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>pA_A</code></td>
<td>
<p>haplotype frequency of AA, the first locus is A and the
second locus is A</p>
</td></tr>
<tr valign="top"><td><code>pA_B</code></td>
<td>
<p>haplotype frequency of AB, the first locus is A and the
second locus is B</p>
</td></tr>
<tr valign="top"><td><code>pB_A</code></td>
<td>
<p>haplotype frequency of BA, the first locus is B and the
second locus is A</p>
</td></tr>
<tr valign="top"><td><code>pB_B</code></td>
<td>
<p>haplotype frequency of BB, the first locus is B and the
second locus is B</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Weir B: Inferences about linkage disequilibrium.
Biometrics 1979; 35: 235-254.
</p>
<p>Weir B: Genetic Data Analysis II. Sunderland, MA: Sinauer Associates, 1996.
</p>
<p>Weir BS, Cockerham CC:
Complete characterization of disequilibrium at two loci;
in Feldman MW (ed): Mathematical Evolutionary Theory.
Princeton, NJ: Princeton University Press, 1989.
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsLDMat.html">snpgdsLDMat</a></code>, <code><a href="snpgdsLDpruning.html">snpgdsLDpruning</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

snp1 &lt;- read.gdsn(index.gdsn(genofile, "genotype"), start=c(1,1), count=c(1,-1))
snp2 &lt;- read.gdsn(index.gdsn(genofile, "genotype"), start=c(2,1), count=c(1,-1))

snpgdsLDpair(snp1, snp2, method = "composite")
snpgdsLDpair(snp1, snp2, method = "r")
snpgdsLDpair(snp1, snp2, method = "dprime")
snpgdsLDpair(snp1, snp2, method = "corr")

# close the genotype file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.10.2 <a href="00Index.html">Index</a>]</div>
</body></html>
