<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Identity-By-State (IBS) proportion</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsIBS {SNPRelate}"><tr><td>snpgdsIBS {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Identity-By-State (IBS) proportion
</h2>

<h3>Description</h3>

<p>Calculate the fraction of identity by state for each pair of samples
</p>


<h3>Usage</h3>

<pre>
snpgdsIBS(gdsobj, sample.id = NULL, snp.id = NULL, autosome.only = TRUE,
    remove.monosnp = TRUE, maf = NaN, missing.rate = NaN,
    num.thread = 1, verbose = TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsobj</code></td>
<td>
<p>an object of class <code><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></code>,
a SNP GDS file</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>a vector of sample id specifying selected samples;
if NULL, all samples are used</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>a vector of snp id specifying selected SNPs; if NULL,
all SNPs are used</p>
</td></tr>
<tr valign="top"><td><code>autosome.only</code></td>
<td>
<p>if <code>TRUE</code>, use autosomal SNPs only; if it is a
numeric or character value, keep SNPs according to the specified
chromosome</p>
</td></tr>
<tr valign="top"><td><code>remove.monosnp</code></td>
<td>
<p>if TRUE, remove monomorphic SNPs</p>
</td></tr>
<tr valign="top"><td><code>maf</code></td>
<td>
<p>to use the SNPs with &quot;&gt;= maf&quot; only; if NaN, no MAF threshold</p>
</td></tr>
<tr valign="top"><td><code>missing.rate</code></td>
<td>
<p>to use the SNPs with &quot;&lt;= missing.rate&quot; only;
if NaN, no missing threshold</p>
</td></tr>
<tr valign="top"><td><code>num.thread</code></td>
<td>
<p>the number of (CPU) cores used; if <code>NA</code>, detect
the number of cores automatically</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The minor allele frequency and missing rate for each SNP passed in
<code>snp.id</code> are calculated over all the samples in <code>sample.id</code>.
</p>
<p>The values of the IBS matrix range from ZERO to ONE.
</p>


<h3>Value</h3>

<p>Return a list (class &quot;snpgdsIBSClass&quot;):
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>the sample ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>the SNP ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>ibs</code></td>
<td>
<p>a matrix of IBS proportion, &quot;# of samples&quot; x &quot;# of samples&quot;</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="snpgdsIBSNum.html">snpgdsIBSNum</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

# perform identity-by-state calculations
ibs &lt;- snpgdsIBS(genofile)

# perform multidimensional scaling analysis on
# the genome-wide IBS pairwise distances:
loc &lt;- cmdscale(1 - ibs$ibs, k = 2)
x &lt;- loc[, 1]; y &lt;- loc[, 2]
race &lt;- as.factor(read.gdsn(index.gdsn(genofile, "sample.annot/pop.group")))
plot(x, y, col=race, xlab = "", ylab = "", main = "cmdscale(IBS Distance)")
legend("topleft", legend=levels(race), text.col=1:nlevels(race))

# close the file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.2.0 <a href="00Index.html">Index</a>]</div>
</body></html>
