<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Principal Component Analysis (PCA) on SNP genotype data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsPCA {SNPRelate}"><tr><td>snpgdsPCA {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Principal Component Analysis (PCA) on SNP genotype data
</h2>

<h3>Description</h3>

<p>To calculate the eigenvectors and eigenvalues for principal component
analysis in GWAS.
</p>


<h3>Usage</h3>

<pre>
snpgdsPCA(gdsobj, sample.id=NULL, snp.id=NULL, autosome.only=TRUE,
    remove.monosnp=TRUE, maf=NaN, missing.rate=NaN, eigen.cnt=32,
    num.thread=1L, bayesian=FALSE, need.genmat=FALSE, genmat.only=FALSE,
    eigen.method=c("DSPEVX", "DSPEV"), verbose = TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsobj</code></td>
<td>
<p>an object of class <code><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></code>,
a SNP GDS file</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>a vector of sample id specifying selected samples;
if NULL, all samples are used</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>a vector of snp id specifying selected SNPs; if NULL,
all SNPs are used</p>
</td></tr>
<tr valign="top"><td><code>autosome.only</code></td>
<td>
<p>if <code>TRUE</code>, use autosomal SNPs only; if it is a
numeric or character value, keep SNPs according to the specified
chromosome</p>
</td></tr>
<tr valign="top"><td><code>remove.monosnp</code></td>
<td>
<p>if TRUE, remove monomorphic SNPs</p>
</td></tr>
<tr valign="top"><td><code>maf</code></td>
<td>
<p>to use the SNPs with &quot;&gt;= maf&quot; only; if NaN, no MAF threshold</p>
</td></tr>
<tr valign="top"><td><code>missing.rate</code></td>
<td>
<p>to use the SNPs with &quot;&lt;= missing.rate&quot; only; if NaN,
no missing threshold</p>
</td></tr>
<tr valign="top"><td><code>eigen.cnt</code></td>
<td>
<p>output the number of eigenvectors; if eigen.cnt &lt;= 0, then
return all eigenvectors</p>
</td></tr>
<tr valign="top"><td><code>num.thread</code></td>
<td>
<p>the number of (CPU) cores used; if <code>NA</code>, detect
the number of cores automatically</p>
</td></tr>
<tr valign="top"><td><code>bayesian</code></td>
<td>
<p>if TRUE, use bayesian normalization</p>
</td></tr>
<tr valign="top"><td><code>need.genmat</code></td>
<td>
<p>if TRUE, return the genetic covariance matrix</p>
</td></tr>
<tr valign="top"><td><code>genmat.only</code></td>
<td>
<p>return the genetic covariance matrix only, do not
compute the eigenvalues and eigenvectors</p>
</td></tr>
<tr valign="top"><td><code>eigen.method</code></td>
<td>
<p>&quot;DSPEVX&quot; &ndash; compute the top <code>eigen.cnt</code>
eigenvalues and eigenvectors using LAPACK::DSPEVX; &quot;DSPEV&quot; &ndash; to be
compatible with SNPRelate_1.1.6 or earlier, using LAPACK::DSPEV;
&quot;DSPEVX&quot; is significantly faster than &quot;DSPEV&quot; if only top principal
components are of interest</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The minor allele frequency and missing rate for each SNP passed in
<code>snp.id</code> are calculated over all the samples in <code>sample.id</code>.
</p>


<h3>Value</h3>

<p>Return a <code>snpgdsPCAClass</code> object, and it is a list:
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>the sample ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>the SNP ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>eigenval</code></td>
<td>
<p>eigenvalues</p>
</td></tr>
<tr valign="top"><td><code>eigenvect</code></td>
<td>
<p>eigenvactors, &quot;# of samples&quot; x &quot;eigen.cnt&quot;</p>
</td></tr>
<tr valign="top"><td><code>varprop</code></td>
<td>
<p>variance proportion for each principal component</p>
</td></tr>
<tr valign="top"><td><code>TraceXTX</code></td>
<td>
<p>the trace of the genetic covariance matrix</p>
</td></tr>
<tr valign="top"><td><code>Bayesian</code></td>
<td>
<p>whether use bayerisan normalization</p>
</td></tr>
<tr valign="top"><td><code>genmat</code></td>
<td>
<p>the genetic covariance matrix</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Patterson, N., Price, A. L. &amp; Reich, D.
Population structure and eigenanalysis. PLoS Genet. 2, e190 (2006).
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsPCACorr.html">snpgdsPCACorr</a></code>, <code><a href="snpgdsPCASampLoading.html">snpgdsPCASampLoading</a></code>,
<code><a href="snpgdsPCASNPLoading.html">snpgdsPCASNPLoading</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

# run PCA
RV &lt;- snpgdsPCA(genofile)

# eigenvalues
head(RV$eigenval)

# variance proportion (%)
head(round(RV$varprop*100, 2))
# [1] 12.23  5.84  1.01  0.95  0.84  0.74


####  there is no population information  ####

# make a data.frame
tab &lt;- data.frame(sample.id = RV$sample.id,
    EV1 = RV$eigenvect[,1],    # the first eigenvector
    EV2 = RV$eigenvect[,2],    # the second eigenvector
    stringsAsFactors = FALSE)
head(tab)
#   sample.id         EV1         EV2
# 1   NA19152 -0.08411287 -0.01226860
# 2   NA19139 -0.08360644 -0.01085849
# 3   NA18912 -0.08110808 -0.01184524
# 4   NA19160 -0.08680864 -0.01447106
# 5   NA07034  0.03109761  0.07709255
# 6   NA07055  0.03228450  0.08155730

# draw
plot(tab$EV2, tab$EV1, xlab="eigenvector 2", ylab="eigenvector 1")



####  there are population information  ####

# get population information
#   or pop_code &lt;- scan("pop.txt", what=character())
#   if it is stored in a text file "pop.txt"
pop_code &lt;- read.gdsn(index.gdsn(genofile, "sample.annot/pop.group"))

# get sample id
samp.id &lt;- read.gdsn(index.gdsn(genofile, "sample.id"))

# assume the order of sample IDs is as the same as population codes
cbind(samp.id, pop_code)
#        samp.id       pop_code
#   [1,] "NA19152"     "YRI"   
#   [2,] "NA19139"     "YRI"   
#   [3,] "NA18912"     "YRI"   
#   [4,] "NA19160"     "YRI"   
#   [5,] "NA07034"     "CEU"   
#   ...

# make a data.frame
tab &lt;- data.frame(sample.id = RV$sample.id,
    pop = factor(pop_code)[match(RV$sample.id, samp.id)],
    EV1 = RV$eigenvect[,1],    # the first eigenvector
    EV2 = RV$eigenvect[,2],    # the second eigenvector
    stringsAsFactors = FALSE)
head(tab)
#   sample.id pop         EV1         EV2
# 1   NA19152 YRI -0.08411287 -0.01226860
# 2   NA19139 YRI -0.08360644 -0.01085849
# 3   NA18912 YRI -0.08110808 -0.01184524
# 4   NA19160 YRI -0.08680864 -0.01447106
# 5   NA07034 CEU  0.03109761  0.07709255
# 6   NA07055 CEU  0.03228450  0.08155730

# draw
plot(tab$EV2, tab$EV1, col=as.integer(tab$pop),
    xlab="eigenvector 2", ylab="eigenvector 1")
legend("bottomright", legend=levels(tab$pop), pch="o", col=1:4)


# close the file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.2.0 <a href="00Index.html">Index</a>]</div>
</body></html>
