<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Create a SNP genotype dataset from a matrix</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsCreateGeno {SNPRelate}"><tr><td>snpgdsCreateGeno {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Create a SNP genotype dataset from a matrix
</h2>

<h3>Description</h3>

<p>To create a GDS file of genotypes from a matrix.
</p>


<h3>Usage</h3>

<pre>
snpgdsCreateGeno(gds.fn, genmat, sample.id=NULL, snp.id=NULL, snp.rs.id=NULL,
    snp.chromosome=NULL, snp.position=NULL, snp.allele=NULL, snpfirstdim=TRUE,
    compress.annotation="ZIP_RA.max", compress.geno="", other.vars=NULL)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gds.fn</code></td>
<td>
<p>the file name of gds</p>
</td></tr>
<tr valign="top"><td><code>genmat</code></td>
<td>
<p>a matrix of genotypes</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>the sample ids, which should be unique</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>the SNP ids, which should be unique</p>
</td></tr>
<tr valign="top"><td><code>snp.rs.id</code></td>
<td>
<p>the rs ids for SNPs, which can be not unique</p>
</td></tr>
<tr valign="top"><td><code>snp.chromosome</code></td>
<td>
<p>the chromosome indices</p>
</td></tr>
<tr valign="top"><td><code>snp.position</code></td>
<td>
<p>the SNP positions in basepair</p>
</td></tr>
<tr valign="top"><td><code>snp.allele</code></td>
<td>
<p>the reference/non-reference alleles</p>
</td></tr>
<tr valign="top"><td><code>snpfirstdim</code></td>
<td>
<p>if TRUE, genotypes are stored in the individual-major
mode, (i.e, list all SNPs for the first individual, and then list
all SNPs for the second individual, etc)</p>
</td></tr>
<tr valign="top"><td><code>compress.annotation</code></td>
<td>
<p>the compression method for the variables except
<code>genotype</code></p>
</td></tr>
<tr valign="top"><td><code>compress.geno</code></td>
<td>
<p>the compression method for the variable
<code>genotype</code></p>
</td></tr>
<tr valign="top"><td><code>other.vars</code></td>
<td>
<p>a list object storing other variables</p>
</td></tr>
</table>


<h3>Details</h3>

<p>There are possible values stored in the variable <code>genmat</code>: 0, 1, 2
and other values. &ldquo;0&rdquo; indicates two B alleles, &ldquo;1&rdquo; indicates one A allele
and one B allele, &ldquo;2&rdquo; indicates two A alleles, and other values indicate a
missing genotype.
</p>
<p>If <code>snpfirstdim</code> is <code>TRUE</code>, then <code>genmat</code> should be &ldquo;# of
SNPs X # of samples&rdquo;; if <code>snpfirstdim</code> is <code>FALSE</code>, then
<code>genmat</code> should be &ldquo;# of samples X # of SNPs&rdquo;.
</p>
<p>The typical variables specified in <code>other.vars</code> are &ldquo;sample.annot&rdquo;
and &ldquo;snp.annot&rdquo;, which are data.frame objects.
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="snpgdsCreateGenoSet.html">snpgdsCreateGenoSet</a></code>, <code><a href="snpgdsCombineGeno.html">snpgdsCombineGeno</a></code>
</p>


<h3>Examples</h3>

<pre>
# load data
data(hapmap_geno)

# create a gds file
with(hapmap_geno, snpgdsCreateGeno("test.gds", genmat=genotype,
    sample.id=sample.id, snp.id=snp.id, snp.chromosome=snp.chromosome,
    snp.position=snp.position, snp.allele=snp.allele, snpfirstdim=TRUE))

# open the gds file
genofile &lt;- snpgdsOpen("test.gds")

RV &lt;- snpgdsPCA(genofile)
plot(RV$eigenvect[,2], RV$eigenvect[,1], xlab="PC 2", ylab="PC 1")

# close the file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.2.0 <a href="00Index.html">Index</a>]</div>
</body></html>
