<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Individual Inbreeding Coefficient</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsIndInbCoef {SNPRelate}"><tr><td>snpgdsIndInbCoef {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Individual Inbreeding Coefficient
</h2>

<h3>Description</h3>

<p>To calculate an individual inbreeding coefficient using SNP genotype data
</p>


<h3>Usage</h3>

<pre>
snpgdsIndInbCoef(x, p, method = c("mom.weir", "mom.visscher", "mle"),
    reltol=.Machine$double.eps^0.75)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>SNP genotypes</p>
</td></tr>
<tr valign="top"><td><code>p</code></td>
<td>
<p>allele frequencies</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
<p>see details</p>
</td></tr>
<tr valign="top"><td><code>reltol</code></td>
<td>
<p>relative convergence tolerance used in MLE; the algorithm
stops if it is unable to reduce the value of log likelihood by a
factor of $reltol * (abs(log likelihood with the initial parameters)
+ reltol)$ at a step.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The <code>method</code> can be:
<code>"mom.weir"</code>: a modified Visscher's estimator, proposed by Bruce Weir;
<code>"mom.visscher"</code>: Visscher's estimator described in Yang et al. (2010);
<code>"mle"</code>: the maximum likelihood estimation.
</p>


<h3>Value</h3>

<p>Return estimated inbreeding coefficient.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Yang J, Benyamin B, McEvoy BP, Gordon S, Henders AK, Nyholt DR, Madden PA,
Heath AC, Martin NG, Montgomery GW, Goddard ME, Visscher PM. 2010.
Common SNPs explain a large proportion of the heritability for human
height. Nat Genet. 42(7):565-9. Epub 2010 Jun 20.
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

chr1 &lt;- read.gdsn(index.gdsn(genofile, "snp.id"))[
    read.gdsn(index.gdsn(genofile, "snp.chromosome"))==1]
chr1idx &lt;- match(chr1, read.gdsn(index.gdsn(genofile, "snp.id")))

AF &lt;- snpgdsSNPRateFreq(genofile)
g &lt;- read.gdsn(index.gdsn(genofile, "genotype"), start=c(1,1), count=c(-1,1))

snpgdsIndInbCoef(g[chr1idx], AF$AlleleFreq[chr1idx], method="mom.weir")
snpgdsIndInbCoef(g[chr1idx], AF$AlleleFreq[chr1idx], method="mom.visscher")
snpgdsIndInbCoef(g[chr1idx], AF$AlleleFreq[chr1idx], method="mle")

# close the genotype file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.2.0 <a href="00Index.html">Index</a>]</div>
</body></html>
