<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Sliding window</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsSlidingWindow {SNPRelate}"><tr><td>snpgdsSlidingWindow {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Sliding window
</h2>

<h3>Description</h3>

<p>Apply a user-defined function with a sliding window.
</p>


<h3>Usage</h3>

<pre>
snpgdsSlidingWindow(gdsobj, sample.id=NULL, snp.id=NULL,
    FUN=NULL, winsize=100000L, shift=10000L, unit=c("basepair", "locus"),
    winstart=NULL, autosome.only=FALSE, remove.monosnp=TRUE, maf=NaN,
    missing.rate=NaN, as.is=c("list", "numeric", "array"),
    with.id=c("snp.id", "snp.id.in.window", "none"), num.thread=1,
    verbose=TRUE, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsobj</code></td>
<td>
<p>an object of class <code><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></code>,
a SNP GDS file</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>a vector of sample id specifying selected samples;
if NULL, all samples are used</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>a vector of snp id specifying selected SNPs; if NULL,
all SNPs are used</p>
</td></tr>
<tr valign="top"><td><code>FUN</code></td>
<td>
<p>the user-defined function, see details</p>
</td></tr>
<tr valign="top"><td><code>winsize</code></td>
<td>
<p>the size of sliding window</p>
</td></tr>
<tr valign="top"><td><code>shift</code></td>
<td>
<p>the amount of shifting the sliding window</p>
</td></tr>
<tr valign="top"><td><code>unit</code></td>
<td>
<p><code>"basepair"</code> &ndash; <code>winsize</code> and <code>shift</code> are
applied with SNP coordinate of basepair; <code>"locus"</code> &ndash;
<code>winsize</code> and <code>shift</code> are applied according to the SNP order
in the GDS file</p>
</td></tr>
<tr valign="top"><td><code>winstart</code></td>
<td>
<p><code>NULL</code> &ndash; no specific starting position; an integer
&ndash; a starting position for all chromosomes; or a vector of integer
&ndash; the starting positions for each chromosome</p>
</td></tr>
<tr valign="top"><td><code>autosome.only</code></td>
<td>
<p>if <code>TRUE</code>, use autosomal SNPs only; if it is a
numeric or character value, keep SNPs according to the specified
chromosome</p>
</td></tr>
<tr valign="top"><td><code>remove.monosnp</code></td>
<td>
<p>if TRUE, remove monomorphic SNPs</p>
</td></tr>
<tr valign="top"><td><code>maf</code></td>
<td>
<p>to use the SNPs with &quot;&gt;= maf&quot; only; if NaN, no MAF threshold</p>
</td></tr>
<tr valign="top"><td><code>missing.rate</code></td>
<td>
<p>to use the SNPs with &quot;&lt;= missing.rate&quot; only; if NaN,
no missing threshold</p>
</td></tr>
<tr valign="top"><td><code>as.is</code></td>
<td>
<p>save the value returned from <code>FUN</code> as &quot;list&quot; or
&quot;numeric&quot;; &quot;array&quot; is equivalent to &quot;numeric&quot; except some cases,
see details</p>
</td></tr>
<tr valign="top"><td><code>with.id</code></td>
<td>
<p>&quot;snp.id&quot;, &quot;snp.id.in.window&quot; or &quot;none&quot;</p>
</td></tr>
<tr valign="top"><td><code>num.thread</code></td>
<td>
<p>the number of (CPU) cores used; if <code>NA</code>, detect
the number of cores automatically</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, show information</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>optional arguments to <code>FUN</code></p>
</td></tr>
</table>


<h3>Details</h3>

<p>If <code>FUN="snpgdsFst"</code>, two additional arguments <code>"population"</code> and
<code>"method"</code> should be specified. <code>"population"</code> and <code>"method"</code>
are defined in <code><a href="snpgdsFst.html">snpgdsFst</a></code>. <code>"as.is"</code> could be &quot;list&quot; (returns
a list of the values from <code><a href="snpgdsFst.html">snpgdsFst</a></code>), &quot;numeric&quot; (
population-average Fst, returns a vector) or &quot;array&quot; (population-average and
-specific Fst, returns a &lsquo;# of pop + 1&rsquo;-by-&lsquo;# of windows&rsquo; matrix, and the first
row is population-average Fst).
</p>
<p>If <code>FUN="snpgdsSNPRateFreq"</code>, .
</p>


<h3>Value</h3>

<p>Return a list
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

# sliding windows
rv &lt;- snpgdsSlidingWindow(genofile, winsize=500000, shift=100000,
    FUN=function(...) NULL)

# plot
plot(rv$chr1.num, ylab="# of SNPs in the sliding window")

# close the genotype file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.2.0 <a href="00Index.html">Index</a>]</div>
</body></html>
