<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Log likelihood for MLE method in the Identity-By-Descent...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsIBDMLELogLik {SNPRelate}"><tr><td>snpgdsIBDMLELogLik {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Log likelihood for MLE method in the Identity-By-Descent (IBD) Analysis
</h2>

<h3>Description</h3>

<p>Calculate the log likelihood values from maximum likelihood estimation.
</p>


<h3>Usage</h3>

<pre>
snpgdsIBDMLELogLik(gdsobj, ibdobj, k0 = NaN, k1 = NaN,
    relatedness=c("", "self", "fullsib", "offspring",
    "halfsib", "cousin", "unrelated"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsobj</code></td>
<td>
<p>an object of class <code><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></code>,
a SNP GDS file</p>
</td></tr>
<tr valign="top"><td><code>ibdobj</code></td>
<td>
<p>the <code>snpgdsIBDClass</code> object returned from
<a href="snpgdsIBDMLE.html">snpgdsIBDMLE</a></p>
</td></tr>
<tr valign="top"><td><code>k0</code></td>
<td>
<p>specified IBD coefficient</p>
</td></tr>
<tr valign="top"><td><code>k1</code></td>
<td>
<p>specified IBD coefficient</p>
</td></tr>
<tr valign="top"><td><code>relatedness</code></td>
<td>
<p>specify a relatedness, otherwise use the values of
k0 and k1</p>
</td></tr>
</table>


<h3>Details</h3>

<p>If (<code>relatedness</code> == &quot;&quot;) and (k0 == NaN or k1 == NaN), then return
the log likelihood values for each (k0, k1) stored in ibdobj. \
If (<code>relatedness</code> == &quot;&quot;) and (k0 != NaN) and (k1 != NaN), then return
the log likelihood values for a specific IBD coefficient (k0, k1). \
If <code>relatedness</code> is: &quot;self&quot;, then k0 = 0, k1 = 0; &quot;fullsib&quot;, then
k0 = 0.25, k1 = 0.5; &quot;offspring&quot;, then k0 = 0, k1 = 1; &quot;halfsib&quot;, then
k0 = 0.5, k1 = 0.5; &quot;cousin&quot;, then k0 = 0.75, k1 = 0.25; &quot;unrelated&quot;, then
k0 = 1, k1 = 0.
</p>


<h3>Value</h3>

<p>Return a n-by-n matrix of log likelihood values, where n is the
number of samples.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Milligan BG. 2003. Maximum-likelihood estimation of relatedness.
Genetics 163:1153-1167.
</p>
<p>Weir BS, Anderson AD, Hepler AB. 2006.
Genetic relatedness analysis: modern data and new challenges.
Nat Rev Genet. 7(10):771-80.
</p>
<p>Choi Y, Wijsman EM, Weir BS. 2009.
Case-control association testing in the presence of unknown
relationships.
Genet Epidemiol 33(8):668-78.
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsIBDMLE.html">snpgdsIBDMLE</a></code>, <code><a href="snpgdsIBDMoM.html">snpgdsIBDMoM</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

YRI.id &lt;- read.gdsn(index.gdsn(genofile, "sample.id"))[
    read.gdsn(index.gdsn(genofile, "sample.annot/pop.group"))=="YRI"]
YRI.id &lt;- YRI.id[1:30]

# SNP pruning
set.seed(10)
snpset &lt;- snpgdsLDpruning(genofile, sample.id=YRI.id, maf=0.05,
    missing.rate=0.05)
snpset &lt;- sample(unlist(snpset), 250)
mibd &lt;- snpgdsIBDMLE(genofile, sample.id=YRI.id, snp.id=snpset)
names(mibd)

# select a set of pairs of individuals
d &lt;- snpgdsIBDSelection(mibd, kinship.cutoff=1/8)
head(d)


# log likelihood

loglik &lt;- snpgdsIBDMLELogLik(genofile, mibd)
loglik0 &lt;- snpgdsIBDMLELogLik(genofile, mibd, relatedness="unrelated")

# likelihood ratio test
p.value &lt;- pchisq(loglik - loglik0, 1, lower.tail=FALSE)


flag &lt;- lower.tri(mibd$k0)
plot(NaN, xlim=c(0,1), ylim=c(0,1), xlab="k0", ylab="k1")
lines(c(0,1), c(1,0), col="red", lty=3)
points(mibd$k0[flag], mibd$k1[flag])

# specify the allele frequencies
afreq &lt;- snpgdsSNPRateFreq(genofile, sample.id=YRI.id,
    snp.id=snpset)$AlleleFreq
subibd &lt;- snpgdsIBDMLE(genofile, sample.id=YRI.id[1:25], snp.id=snpset,
    allele.freq=afreq)
summary(c(subibd$k0 - mibd$k0[1:25, 1:25]))
# ZERO
summary(c(subibd$k1 - mibd$k1[1:25, 1:25]))
# ZERO


# close the genotype file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.10.2 <a href="00Index.html">Index</a>]</div>
</body></html>
