<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Conversion from GDS to Eigen (EIGENSTRAT)</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsGDS2Eigen {SNPRelate}"><tr><td>snpgdsGDS2Eigen {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Conversion from GDS to Eigen (EIGENSTRAT)</h2>

<h3>Description</h3>

<p>Convert a GDS file to an EIGENSTRAT file.
</p>


<h3>Usage</h3>

<pre>
snpgdsGDS2Eigen(gdsobj, eigen.fn, sample.id=NULL, snp.id=NULL, verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsobj</code></td>
<td>
<p>an object of class <code><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></code>,
a SNP GDS file</p>
</td></tr>
<tr valign="top"><td><code>eigen.fn</code></td>
<td>
<p>the file name of EIGENSTRAT</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>a vector of sample id specifying selected samples;
if NULL, all samples are used</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>a vector of snp id specifying selected SNPs; if NULL,
all SNPs are used</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>GDS &ndash; Genomic Data Structures, the extended file name used for storing
genetic data, and the file format used in the <a href="../../gdsfmt/html/gdsfmt-package.html">gdsfmt</a> package.
</p>
<p>Eigen &ndash; the text format used in EIGENSTRAT.
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Patterson N, Price AL, Reich D (2006) Population structure and
eigenanalysis. PLoS Genetics 2:e190.
</p>
<p>Price AL, Patterson NJ, Plenge RM, Weinblatt ME, Shadick NA, Reich D (2006)
Principal components analysis corrects for stratification in
genome-wide association studies.
Nat Genet. 38, 904-909.
</p>
<p><a href="http://corearray.sourceforge.net/">http://corearray.sourceforge.net/</a>
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsGDS2PED.html">snpgdsGDS2PED</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

snpset &lt;- snpgdsSelectSNP(genofile, missing.rate=0.95)
snpgdsGDS2Eigen(genofile, eigen.fn="tmpeigen", snp.id=snpset)

# close the genotype file
snpgdsClose(genofile)


# delete the temporary files
unlink(c("tmpeigen.eigenstratgeno", "tmpeigen.ind", "tmpeigen.snp"), force=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.10.2 <a href="00Index.html">Index</a>]</div>
</body></html>
