<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Estimate ancestral proportions from the eigen-analysis</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsAdmixProp {SNPRelate}"><tr><td>snpgdsAdmixProp {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Estimate ancestral proportions from the eigen-analysis
</h2>

<h3>Description</h3>

<p>Estimate ancestral (admixture) proportions based on the eigen-analysis.
</p>


<h3>Usage</h3>

<pre>
snpgdsAdmixProp(eigobj, groups, bound=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>eigobj</code></td>
<td>
<p>an object of <code>snpgdsEigMixClass</code> from
<code><a href="snpgdsEIGMIX.html">snpgdsEIGMIX</a></code>, or an object of <code>snpgdsPCAClass</code>
from <code><a href="snpgdsPCA.html">snpgdsPCA</a></code></p>
</td></tr>
<tr valign="top"><td><code>groups</code></td>
<td>
<p>a list of sample IDs, such like <code>groups = list(
        CEU = c("NA0101", "NA1022", ...), YRI = c("NAxxxx", ...),
        Asia = c("NA1234", ...))</code></p>
</td></tr>
<tr valign="top"><td><code>bound</code></td>
<td>
<p>if <code>TRUE</code>, the estimates are bounded so that no
component &lt; 0 or &gt; 1, and the sum of proportions is one</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The minor allele frequency and missing rate for each SNP passed in
<code>snp.id</code> are calculated over all the samples in <code>sample.id</code>.
</p>


<h3>Value</h3>

<p>Return a <code>snpgdsEigMixClass</code> object, and it is a list:
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>the sample ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>snp.id</code></td>
<td>
<p>the SNP ids used in the analysis</p>
</td></tr>
<tr valign="top"><td><code>eigenval</code></td>
<td>
<p>eigenvalues</p>
</td></tr>
<tr valign="top"><td><code>eigenvect</code></td>
<td>
<p>eigenvactors, &quot;# of samples&quot; x &quot;eigen.cnt&quot;</p>
</td></tr>
<tr valign="top"><td><code>ibdmat</code></td>
<td>
<p>the IBD matrix</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Zheng X, Weir BS.
Eigenanalysis on SNP Data with an Interpretation of Identity by Descent.
Theoretical Population Biology. 2015 Oct 23. pii: S0040-5809(15)00089-1.
doi: 10.1016/j.tpb.2015.09.004. [Epub ahead of print]
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsEIGMIX.html">snpgdsEIGMIX</a></code>, <code><a href="snpgdsPCA.html">snpgdsPCA</a></code>,
<code><a href="snpgdsAdmixPlot.html">snpgdsAdmixPlot</a></code>
</p>


<h3>Examples</h3>

<pre>
# open an example dataset (HapMap)
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

# get population information
#   or pop_code &lt;- scan("pop.txt", what=character())
#   if it is stored in a text file "pop.txt"
pop_code &lt;- read.gdsn(index.gdsn(genofile, "sample.annot/pop.group"))

# get sample id
samp.id &lt;- read.gdsn(index.gdsn(genofile, "sample.id"))

# run eigen-analysis
RV &lt;- snpgdsEIGMIX(genofile)

# eigenvalues
RV$eigenval

# make a data.frame
tab &lt;- data.frame(sample.id = samp.id, pop = factor(pop_code),
    EV1 = RV$eigenvect[,1],    # the first eigenvector
    EV2 = RV$eigenvect[,2],    # the second eigenvector
    stringsAsFactors = FALSE)
head(tab)

# draw
plot(tab$EV2, tab$EV1, col=as.integer(tab$pop),
    xlab="eigenvector 2", ylab="eigenvector 1")
legend("topleft", legend=levels(tab$pop), pch="o", col=1:4)


# define groups
groups &lt;- list(CEU = samp.id[pop_code == "CEU"],
    YRI = samp.id[pop_code == "YRI"],
    CHB = samp.id[is.element(pop_code, c("HCB", "JPT"))])

prop &lt;- snpgdsAdmixProp(RV, groups=groups)

# draw
plot(prop[, "YRI"], prop[, "CEU"], col=as.integer(tab$pop),
    xlab = "Admixture Proportion from YRI",
    ylab = "Admixture Proportion from CEU")
abline(v=0, col="gray25", lty=2)
abline(h=0, col="gray25", lty=2)
abline(a=1, b=-1, col="gray25", lty=2)
legend("topright", legend=levels(tab$pop), pch="o", col=1:4)


# draw
snpgdsAdmixPlot(prop, group=pop_code)


# close the genotype file
snpgdsClose(genofile)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.10.2 <a href="00Index.html">Index</a>]</div>
</body></html>
