<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Parallel Computing Toolset for Relatedness and Principal
Component Analysis of SNP Data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Parallel Computing Toolset for Relatedness and Principal
Component Analysis of SNP Data
</h1>
<hr/>
<div style="text-align: center;">
</div><h2>Documentation for package &lsquo;SNPRelate&rsquo; version 1.3.5</h2>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="SNPRelate-package.html">SNPRelate-package</a></td>
<td>Parallel Computing Toolset for Genome-Wide Association Studies</td></tr>
<tr><td style="width: 25%;"><a href="hapmap_geno.html">hapmap_geno</a></td>
<td>SNP genotypes of HapMap samples</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsAdmixProp.html">snpgdsAdmixProp</a></td>
<td>Estimate ancestral proportions from the eigen-analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsAlleleSwitch.html">snpgdsAlleleSwitch</a></td>
<td>Allele-switching</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsApartSelection.html">snpgdsApartSelection</a></td>
<td>Select SNPs with a basepair distance</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsBED2GDS.html">snpgdsBED2GDS</a></td>
<td>Conversion from PLINK BED to GDS</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsClose.html">snpgdsClose</a></td>
<td>Close the SNP GDS File</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsCombineGeno.html">snpgdsCombineGeno</a></td>
<td>Merge SNP datasets</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsCreateGeno.html">snpgdsCreateGeno</a></td>
<td>Create a SNP genotype dataset from a matrix</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsCreateGenoSet.html">snpgdsCreateGenoSet</a></td>
<td>Create a SNP genotype dataset from a GDS file</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsCutTree.html">snpgdsCutTree</a></td>
<td>Determine clusters of individuals</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsDiss.html">snpgdsDiss</a></td>
<td>Individual dissimilarity analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsDrawTree.html">snpgdsDrawTree</a></td>
<td>Draw a dendrogram</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsEIGMIX.html">snpgdsEIGMIX</a></td>
<td>Eigen-analysis on SNP genotype data</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsErrMsg.html">snpgdsErrMsg</a></td>
<td>Get the last error information</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsExampleFileName.html">snpgdsExampleFileName</a></td>
<td>Example GDS file</td></tr>
<tr><td style="width: 25%;"><a href="SNPGDSFileClass-class.html">SNPGDSFileClass</a></td>
<td>SNPGDSFileClass</td></tr>
<tr><td style="width: 25%;"><a href="SNPGDSFileClass-class.html">SNPGDSFileClass-class</a></td>
<td>SNPGDSFileClass</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsFst.html">snpgdsFst</a></td>
<td>F-statistics (fixation indices)</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGDS2BED.html">snpgdsGDS2BED</a></td>
<td>Conversion from GDS to PLINK BED</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGDS2Eigen.html">snpgdsGDS2Eigen</a></td>
<td>Conversion from GDS to Eigen (EIGENSTRAT)</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGDS2PED.html">snpgdsGDS2PED</a></td>
<td>Conversion from GDS to PED</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGEN2GDS.html">snpgdsGEN2GDS</a></td>
<td>Conversion from Oxford GEN format to GDS</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGetGeno.html">snpgdsGetGeno</a></td>
<td>To get a genotype matrix</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsGRM.html">snpgdsGRM</a></td>
<td>Genetic Relationship Matrix (GRM) for SNP genotype data</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsHCluster.html">snpgdsHCluster</a></td>
<td>Hierarchical cluster analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsHWE.html">snpgdsHWE</a></td>
<td>Statistical test of Hardy-Weinberg Equilibrium</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBDKING.html">snpgdsIBDKING</a></td>
<td>KING method of moment for the identity-by-descent (IBD) analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBDMLE.html">snpgdsIBDMLE</a></td>
<td>Maximum likelihood estimation (MLE) for the Identity-By-Descent (IBD) Analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBDMLELogLik.html">snpgdsIBDMLELogLik</a></td>
<td>Log likelihood for MLE method in the Identity-By-Descent (IBD) Analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBDMoM.html">snpgdsIBDMoM</a></td>
<td>PLINK method of moment (MoM) for the Identity-By-Descent (IBD) Analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBDSelection.html">snpgdsIBDSelection</a></td>
<td>Get a table of IBD coefficients</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBS.html">snpgdsIBS</a></td>
<td>Identity-By-State (IBS) proportion</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIBSNum.html">snpgdsIBSNum</a></td>
<td>Identity-By-State (IBS)</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIndInb.html">snpgdsIndInb</a></td>
<td>Individual Inbreeding Coefficients</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsIndInbCoef.html">snpgdsIndInbCoef</a></td>
<td>Individual Inbreeding Coefficient</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsLDMat.html">snpgdsLDMat</a></td>
<td>Linkage Disequilibrium (LD) analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsLDpair.html">snpgdsLDpair</a></td>
<td>Linkage Disequilibrium (LD)</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsLDpruning.html">snpgdsLDpruning</a></td>
<td>Linkage Disequilibrium (LD) based SNP pruning</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsOpen.html">snpgdsOpen</a></td>
<td>Open a SNP GDS File</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsOption.html">snpgdsOption</a></td>
<td>Option settings: chromosome coding, etc</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPairIBD.html">snpgdsPairIBD</a></td>
<td>Calculate Identity-By-Descent (IBD) Coefficients</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPairIBDMLELogLik.html">snpgdsPairIBDMLELogLik</a></td>
<td>Log likelihood for MLE method in the Identity-By-Descent (IBD) Analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPairScore.html">snpgdsPairScore</a></td>
<td>Genotype Score for Pairs of Individuals</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPCA.html">snpgdsPCA</a></td>
<td>Principal Component Analysis (PCA) on SNP genotype data</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPCACorr.html">snpgdsPCACorr</a></td>
<td>PCA-correlated SNPs in principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPCASampLoading.html">snpgdsPCASampLoading</a></td>
<td>Project individuals onto existing principal component axes</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPCASNPLoading.html">snpgdsPCASNPLoading</a></td>
<td>SNP loadings in principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsPED2GDS.html">snpgdsPED2GDS</a></td>
<td>Conversion from PLINK PED to GDS</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSampMissRate.html">snpgdsSampMissRate</a></td>
<td>Missing Rate of Samples</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSelectSNP.html">snpgdsSelectSNP</a></td>
<td>SNP selection</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSlidingWindow.html">snpgdsSlidingWindow</a></td>
<td>Sliding window</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSNPList.html">snpgdsSNPList</a></td>
<td>Create a SNP list object</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSNPListClass.html">snpgdsSNPListClass</a></td>
<td>the class of a SNP list</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSNPListIntersect.html">snpgdsSNPListIntersect</a></td>
<td>Get a common SNP list between two SNP list objects</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSNPListStrand.html">snpgdsSNPListStrand</a></td>
<td>Switch allele strand.</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSNPRateFreq.html">snpgdsSNPRateFreq</a></td>
<td>Allele Frequency, Minor Allele Frequency, Missing Rate of SNPs</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsSummary.html">snpgdsSummary</a></td>
<td>Summary of GDS genotype file</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsTranspose.html">snpgdsTranspose</a></td>
<td>Transpose genotypic matrix</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsVCF2GDS.html">snpgdsVCF2GDS</a></td>
<td>Reformat VCF file(s)</td></tr>
<tr><td style="width: 25%;"><a href="snpgdsVCF2GDS_R.html">snpgdsVCF2GDS_R</a></td>
<td>Reformat a VCF file (R implementation)</td></tr>
<tr><td style="width: 25%;"><a href="SNPRelate-package.html">SNPRelate</a></td>
<td>Parallel Computing Toolset for Genome-Wide Association Studies</td></tr>
</table>
</body></html>
