<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Conversion from PLINK PED to GDS</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpgdsPED2GDS {SNPRelate}"><tr><td>snpgdsPED2GDS {SNPRelate}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Conversion from PLINK PED to GDS</h2>

<h3>Description</h3>

<p>Convert a PLINK PED text file to a GDS file.
</p>


<h3>Usage</h3>

<pre>
snpgdsPED2GDS(ped.fn, map.fn, out.gdsfn, family=TRUE, snpfirstdim=FALSE,
    compress.annotation="ZIP_RA.max", compress.geno="", verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>ped.fn</code></td>
<td>
<p>the file name of PED file, genotype information</p>
</td></tr>
<tr valign="top"><td><code>map.fn</code></td>
<td>
<p>the file name of MAP file</p>
</td></tr>
<tr valign="top"><td><code>out.gdsfn</code></td>
<td>
<p>the output GDS file</p>
</td></tr>
<tr valign="top"><td><code>family</code></td>
<td>
<p>if <code>TRUE</code>, to include family information in the
sample annotation</p>
</td></tr>
<tr valign="top"><td><code>snpfirstdim</code></td>
<td>
<p>if TRUE, genotypes are stored in the individual-major
mode, (i.e, list all SNPs for the first individual, and then list all
SNPs for the second individual, etc)</p>
</td></tr>
<tr valign="top"><td><code>compress.annotation</code></td>
<td>
<p>the compression method for the GDS variables,
except &quot;genotype&quot;; optional values are defined in the function
<code>add.gdsn</code></p>
</td></tr>
<tr valign="top"><td><code>compress.geno</code></td>
<td>
<p>the compression method for &quot;genotype&quot;; optional
values are defined in the function <code>add.gdsn</code></p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>GDS &ndash; Genomic Data Structures, the extended file name used for storing
genetic data, and the file format is used in the <a href="../../gdsfmt/html/gdsfmt-package.html">gdsfmt</a> package.
</p>
<p>PED &ndash; PLINK PED format.
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D,
Maller J, Sklar P, de Bakker PIW, Daly MJ &amp; Sham PC. 2007.
PLINK: a toolset for whole-genome association and population-based linkage
analysis. American Journal of Human Genetics, 81.
</p>


<h3>See Also</h3>

<p><code><a href="snpgdsGDS2PED.html">snpgdsGDS2PED</a></code>, <code><a href="snpgdsBED2GDS.html">snpgdsBED2GDS</a></code>,
<code><a href="snpgdsGDS2BED.html">snpgdsGDS2BED</a></code>
</p>


<h3>Examples</h3>

<pre>
# open
genofile &lt;- snpgdsOpen(snpgdsExampleFileName())

snpgdsGDS2PED(genofile, "tmp")

# close
snpgdsClose(genofile)


# PED ==&gt; GDS
snpgdsPED2GDS("tmp.ped", "tmp.map", "test.gds")


# delete the temporary file
unlink(c("tmp.ped", "tmp.map", "test.gds"), force=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>SNPRelate</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
